libraries/AP_HAL_ChibiOS/stdio.cpp

```diff
@@ -23,54 +23,64 @@
   without an implementation. If we need them then we can implement as
   needed.
  */
-#include <posix.h>
 #include <string.h>
 #include <hal.h>
 #include <memstreams.h>
 #include <chprintf.h>
 #include <ctype.h>
-#include "stdio.h"
-
-int vsnprintf(char *str, size_t size, const char *fmt, va_list ap)
-{
-  int retval = 0;
-#ifndef HAL_NO_PRINTF
-  MemoryStream ms;
-  BaseSequentialStream *chp;
-  size_t size_wo_nul;
-
-  if (size > 0)
-    size_wo_nul = size - 1;
-  else
-    size_wo_nul = 0;
-
-  /* Memory stream object to be used as a string writer, reserving one
-     byte for the final zero.*/
-  msObjectInit(&ms, (uint8_t *)str, size_wo_nul, 0);
-
-  /* Performing the print operation using the common code.*/
-  chp = (BaseSequentialStream *)(void *)&ms;
-
-  retval = chvprintf(chp, fmt, ap);
+#include "hwdef/common/posix.h"
+#include "hwdef/common/stdio.h"
+#include <AP_HAL/AP_HAL.h>
+#include <AP_HAL/utility/print_vprintf.h>
+
+extern const AP_HAL::HAL& hal;
+
+/* Helper class implements AP_HAL::Print so we can use utility/vprintf */
```

Comment:
This looks generic enough that it could be in other HALs too. I think it would be worth considering moving this to AP_HAL/utility, particularly so it can be used in SITL and we can test it with valgrind.